server/src/test/java/org/elasticsearch/cluster/coordination/CoordinatorTests.java

```diff
@@ -1266,6 +1266,66 @@ public void assertMatched() {
+        }
+        logger.info("--> emitting warnings every [{}ms]", warningDelayMillis);
+
+        try (Cluster cluster = new Cluster(between(3, 5), true, settings)) {
+            cluster.runRandomly();
+            cluster.stabilise();
+
+            // drop the publication messages to one node, but then restore connectivity so it remains in the cluster and does not fail
+            // health checks
+            final ClusterNode brokenNode = cluster.getAnyNodeExcept(cluster.getAnyLeader());
+            brokenNode.blackhole();
+            cluster.getAnyLeader().submitValue(randomLong());
+            cluster.runFor(DEFAULT_CLUSTER_STATE_UPDATE_DELAY, "waiting for publication to start");
+            brokenNode.heal();
+
+            final MockLogAppender mockLogAppenderForDelayedPublication = new MockLogAppender();
+            try {
+                Loggers.addAppender(LogManager.getLogger(Coordinator.CoordinatorPublication.class), mockLogAppenderForDelayedPublication);
+                mockLogAppenderForDelayedPublication.start();
+                mockLogAppenderForDelayedPublication.addExpectation(new MockLogAppender.SeenEventExpectation("warning",
+                    Coordinator.CoordinatorPublication.class.getCanonicalName(), Level.WARN,
+                    "after [*] publication of cluster state version [*] is still waiting for " + brokenNode.getLocalNode() + " ["
+                        + Publication.PublicationTargetState.SENT_PUBLISH_REQUEST + ']'));
+                cluster.runFor(warningDelayMillis + DEFAULT_DELAY_VARIABILITY, "waiting for warning to be emitted");
+                mockLogAppenderForDelayedPublication.assertAllExpectationsMatched();
+            } finally {
+                mockLogAppenderForDelayedPublication.stop();
+                Loggers.removeAppender(LogManager.getLogger(Coordinator.CoordinatorPublication.class),
+                    mockLogAppenderForDelayedPublication);
+            }
+
+            final MockLogAppender mockLogAppenderForLagDetection = new MockLogAppender();
+            try {
+                Loggers.addAppender(LogManager.getLogger(LagDetector.class), mockLogAppenderForLagDetection);
+                mockLogAppenderForLagDetection.start();
+                mockLogAppenderForLagDetection.addExpectation(new MockLogAppender.SeenEventExpectation("warning",
+                    LagDetector.class.getCanonicalName(), Level.WARN,
+                    "node [" + brokenNode + "] is lagging at cluster state version [*], " +
+                        "although publication of cluster state version [*] completed [*] ago"));
+                cluster.runFor(defaultMillis(PUBLISH_TIMEOUT_SETTING)
+                    + defaultMillis(LagDetector.CLUSTER_FOLLOWER_LAG_TIMEOUT_SETTING), "waiting for warning to be emitted");
+                mockLogAppenderForLagDetection.assertAllExpectationsMatched();
+            } finally {
+                mockLogAppenderForLagDetection.stop();
```

Comment:
I think it doesn't matter here, but in general you should always remove appenders and then stop them to not get an error for trying to use a stopped appender? (same comment in the other log appender spots)

Reply:
Ok, thanks, adjusted in c9df59c. It doesn't matter here indeed, but a stop-before-remove in places with more concurrency is effectively an assertion that no more log messages are being emitted.